torch/csrc/jit/ivalue.h

```diff
@@ -0,0 +1,270 @@
+  at::Tensor toTensor() && {
+    JIT_ASSERT(isTensor());
+    at::Tensor t(as_tensor_impl, /*retain=*/false);
+    clearToZero();
+    return t;
+  }
+  at::Tensor toTensor() const & {
+    JIT_ASSERT(isTensor());
+    return at::Tensor(as_tensor_impl, /*retain=*/true);
+  }
+
+  // Tuple
+  IValue(Shared<Tuple> v);
+  bool isTuple() const { return Tag::Tuple == tag; }
+  Shared<Tuple> toTuple() && {
+    JIT_ASSERT(isTuple());
+    return moveToRetainable<Tuple>();
+  }
+  Shared<Tuple> toTuple() const & {
+    JIT_ASSERT(isTuple());
+    return toRetainable<Tuple>();
+  }
+
+  // Double
+  IValue(double d)
+  : tag(Tag::Double), retainable(false) {
+    as_double = d;
+  }
+  bool isDouble() const { return Tag::Double == tag; }
+  double toDouble() const {
+    JIT_ASSERT(isDouble());
+    return as_double;
+  }
+
+  // Int
+  IValue(int64_t i)
+  : tag(Tag::Int), retainable(false) {
+    as_int = i;
+  }
+  // allow you to pass literals (3, 4) without ambiguity
+  IValue(int32_t i)
+  : IValue(static_cast<int64_t>(i)) {}
+
+  bool isInt() const { return Tag::Int == tag; }
+  int64_t toInt() const {
+    JIT_ASSERT(isInt());
+    return as_int;
+  }
+
+  // IntList
+  IValue(Shared<IntList> v);
+  bool isIntList() const { return Tag::IntList == tag; }
+  Shared<IntList> toIntList() && {
+    JIT_ASSERT(isIntList());
+    return moveToRetainable<IntList>();
+  }
+  Shared<IntList> toIntList() const & {
+    JIT_ASSERT(isIntList());
+    return toRetainable<IntList>();
+  }
+
+  // DoubleList
+  IValue(Shared<DoubleList> v);
+  bool isDoubleList() const { return Tag::DoubleList == tag; }
+  Shared<DoubleList> toDoubleList() && {
+    JIT_ASSERT(isDoubleList());
+    return moveToRetainable<DoubleList>();
+  }
+  Shared<DoubleList> toDoubleList() const & {
+    JIT_ASSERT(isDoubleList());
+    return toRetainable<DoubleList>();
+  }
+
+private:
+  template<typename T>
+  Shared<T> moveToRetainable() {
+    Shared<T> t(static_cast<T*>(as_retainable), false);
+    clearToZero();
+    return t;
+  }
+  template<typename T>
+  Shared<T> toRetainable() const {
+    return Shared<T>(static_cast<T*>(as_retainable), true);
+  }
+  void clearToZero() {
+    as_int = 0;
+    tag = Tag::Int;
+    retainable = false;
+  }
+  enum class Tag : uint32_t {
+    Tensor, Double, Int, Tuple, IntList, DoubleList
+  };
+  union {
+    at::TensorImpl* as_tensor_impl;
+    at::Retainable* as_retainable;
+    double as_double;
+    int64_t as_int;
+  };
+  Tag tag;
+  bool retainable;
+};
+
+
+// non-mutable list
+template<typename Elem>
+struct ConstantList : at::Retainable {
+ private:
+  ConstantList(std::vector<Elem> elements_)
+  : elements_(std::move(elements_)) {}
+  std::vector<Elem> elements_;
+ public:
+  static Shared<ConstantList<Elem>> create(std::vector<Elem> elements_) {
+    return Shared<ConstantList<Elem>>(
+        new ConstantList<Elem>(std::move(elements_)), false);
+  }
+  at::ArrayRef<Elem> elements() const {
```

Comment:
It's a bit funny that you can modify the elements of a `ConstantList`. I guess we only care about having a constant length, right?

Reply:
ArrayRef is an immutable view as far as I understand, so I don't believe you can modify the members.